feat: add `util` module

tomyrd opened the pull request:
addresses some of the changes mentioned in #905 

This PR adds a `utils` module with some functions which were only avaliable for testing. The functions are refactored to remove unwraps, improve documentation and usability. Some additional specific changes:
- Adds a new `wait_until` function in the client which can receive an arbitrary condition to wait for. All `wait_...` functions were refactored to use it.
- Any duplicate method in the `testing` module was removed or refactored.
- `discarded_transactions` was added to `SyncSummary` to aid in `wait_for_tx`.
- Some of the tests were refactored to use the new `utils`.

igamigo reviewed (commented):
Looks good! I think we may only want to expose just a subset of these, and keep the rest under `testing`. Another approach would be to say that these are all for `testing` for now and maybe we expose them through a comprehensive struct or something similar later. @partylikeits1983 you might have more input here as you have used the client and have context as the issue creator.
In general though, this is a nice refactor for implementing errors in tests as well.

bobbinth commented:
> @partylikeits1983 you might have more input here as you have used the client and have context as the issue creator.

Without having looked at the PR in detail, I'll defer to @partylikeits1983 and @mmagician. I do think we should probably do a more comprehensive refactor to support this functionality, but I also don't mind releasing it now with the caveat that it should be considered unstable and could go away in the next release (but that applies to pretty much everything we do at this point).

tomyrd commented:
> I think we may only want to expose just a subset of these, and keep the rest under `testing`.

Some of these were already exposed just for testing but I moved them to `utils` because they were part of utility tool list mentioned in #905 .
My understanding was that some of this functionality would make for a simpler developer experience, that's why I moved them out of `testing`.  But I realize that we may want to expose it in a better fashion than just standalone functions.

I'll move them back to `testing` and we can review how we want to expose them in v0.11.
